# 🎯 Gerador de Combinações por Soma - Lotofácil

Este notebook gera todas as combinações possíveis de Lotofácil (15 números de 1-25) que somam um valor específico.

**Recursos:**
- Geração de combinações por soma específica
- Filtros por composição (distribuição em grupos c1-c5)
- Filtros por paridade (pares/ímpares)
- Filtros por similaridade com combinação de referência
- Análise estatística completa

---

## 1. Importar Bibliotecas Necessárias

In [1]:
import pandas as pd
from itertools import combinations
from pathlib import Path
from datetime import datetime
from tqdm import tqdm
from collections import Counter

print("✅ Bibliotecas importadas com sucesso!")

✅ Bibliotecas importadas com sucesso!


## 2. Configurar Parâmetros

Defina a **soma alvo** e o **caminho de saída** do arquivo CSV:

In [154]:
# ========== CONFIGURAÇÃO ==========
SOMA_ALVO = 217  # ⬅️ ALTERE AQUI A SOMA DESEJADA
ULTIMOS_N_JOGOS = 25  # ⬅️ QUANTIDADE DE JOGOS RECENTES PARA ANÁLISE
# ==================================

# Diretório de dados
data_dir = Path('data')
data_dir.mkdir(exist_ok=True)

# Carregar dados históricos
arquivo_historico = data_dir / 'lotofacil_sorteios.csv'

if arquivo_historico.exists():
    df_historico = pd.read_csv(arquivo_historico)
    
    # Pegar últimos N jogos
    df_ultimos = df_historico.tail(ULTIMOS_N_JOGOS).copy()
    
    print(f"📊 Dados históricos carregados: {len(df_historico):,} concursos")
    print(f"📊 Analisando últimos {ULTIMOS_N_JOGOS} jogos (concursos {df_ultimos['Concurso'].min()} a {df_ultimos['Concurso'].max()})")
else:
    df_historico = None
    df_ultimos = None
    print(f"⚠️  Arquivo de histórico não encontrado: {arquivo_historico}")
    print(f"   Análise histórica será ignorada.")

# Nome do arquivo de saída
arquivo_saida = data_dir / f'combinacoes_soma_{SOMA_ALVO}.csv'

print(f"\n📊 Soma Alvo: {SOMA_ALVO}")
print(f"📁 Arquivo de saída: {arquivo_saida}")

📊 Dados históricos carregados: 3,589 concursos
📊 Analisando últimos 25 jogos (concursos 3565 a 3589)

📊 Soma Alvo: 217
📁 Arquivo de saída: data\combinacoes_soma_217.csv


## 3. Gerar Todas as Combinações com a Soma Especificada

Este processo pode levar alguns minutos dependendo da soma escolhida.

In [155]:
print(f"Gerando combinações de 15 números (1-25) com soma = {SOMA_ALVO}...")
print("Por favor, aguarde. Isso pode levar alguns minutos.\n")

# Lista para armazenar as combinações
combinacoes_validas = []

# Total de combinações possíveis C(25, 15) = 3,268,760
total_combinacoes = 3268760

# Gerar todas as combinações de 15 números
inicio = datetime.now()

for combo in tqdm(combinations(range(1, 26), 15), total=total_combinacoes, desc="Processando"):
    if sum(combo) == SOMA_ALVO:
        combinacoes_validas.append(combo)

fim = datetime.now()
duracao = (fim - inicio).total_seconds()

print(f"\n{'='*70}")
print(f"✅ Processamento concluído!")
print(f"{'='*70}")
print(f"⏱️  Tempo de processamento: {duracao:.2f} segundos")
print(f"📊 Total de combinações encontradas: {len(combinacoes_validas):,}")
print(f"{'='*70}")

Gerando combinações de 15 números (1-25) com soma = 217...
Por favor, aguarde. Isso pode levar alguns minutos.



Processando: 100%|██████████| 3268760/3268760 [00:01<00:00, 1795948.41it/s]


✅ Processamento concluído!
⏱️  Tempo de processamento: 1.82 segundos
📊 Total de combinações encontradas: 35,178


## 4. Criar DataFrame e Salvar em CSV

In [156]:
# Criar DataFrame com as combinações
dados = []

for i, combo in enumerate(combinacoes_validas, 1):
    # Formatar combinação como string "01-02-03-..."
    combo_str = '-'.join([f'{num:02d}' for num in combo])
    
    dados.append({
        'ID': i,
        'Soma': SOMA_ALVO,
        'Combinacao': combo_str,
        'Numeros': str(list(combo))
    })

df_combinacoes = pd.DataFrame(dados)

# Salvar em CSV
df_combinacoes.to_csv(arquivo_saida, index=False, encoding='utf-8-sig')

print(f"\n✅ Arquivo salvo com sucesso!")
print(f"📁 {arquivo_saida}")
print(f"\n📊 Primeiras 10 combinações:")
print(df_combinacoes.head(10).to_string(index=False))


✅ Arquivo salvo com sucesso!
📁 data\combinacoes_soma_217.csv

📊 Primeiras 10 combinações:
 ID  Soma                                   Combinacao                                                 Numeros
  1   217 01-02-03-04-05-13-17-18-19-20-21-22-23-24-25 [1, 2, 3, 4, 5, 13, 17, 18, 19, 20, 21, 22, 23, 24, 25]
  2   217 01-02-03-04-05-14-16-18-19-20-21-22-23-24-25 [1, 2, 3, 4, 5, 14, 16, 18, 19, 20, 21, 22, 23, 24, 25]
  3   217 01-02-03-04-05-15-16-17-19-20-21-22-23-24-25 [1, 2, 3, 4, 5, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25]
  4   217 01-02-03-04-06-12-17-18-19-20-21-22-23-24-25 [1, 2, 3, 4, 6, 12, 17, 18, 19, 20, 21, 22, 23, 24, 25]
  5   217 01-02-03-04-06-13-16-18-19-20-21-22-23-24-25 [1, 2, 3, 4, 6, 13, 16, 18, 19, 20, 21, 22, 23, 24, 25]
  6   217 01-02-03-04-06-14-15-18-19-20-21-22-23-24-25 [1, 2, 3, 4, 6, 14, 15, 18, 19, 20, 21, 22, 23, 24, 25]
  7   217 01-02-03-04-06-14-16-17-19-20-21-22-23-24-25 [1, 2, 3, 4, 6, 14, 16, 17, 19, 20, 21, 22, 23, 24, 25]
  8   217 01-02-03-04

## 5. Análise Estatística das Combinações Geradas

In [157]:
print(f"\n{'='*70}")
print(f"ESTATÍSTICAS DAS COMBINAÇÕES - SOMA {SOMA_ALVO}")
print(f"{'='*70}\n")

# Análise de paridade
pares_por_combinacao = []
impares_por_combinacao = []

for combo in combinacoes_validas:
    pares = sum(1 for n in combo if n % 2 == 0)
    impares = 15 - pares
    pares_por_combinacao.append(pares)
    impares_por_combinacao.append(impares)

# Contar distribuições
from collections import Counter
distribuicao_pares = Counter(pares_por_combinacao)

print("📊 Distribuição de Paridade:")
print("-" * 70)
for pares in sorted(distribuicao_pares.keys()):
    qtd = distribuicao_pares[pares]
    pct = (qtd / len(combinacoes_validas)) * 100
    impares = 15 - pares
    bar = "█" * int(pct / 2)
    print(f"  {pares} pares / {impares} ímpares: {qtd:6,} ({pct:5.2f}%) {bar}")

# Análise de números primos
print(f"\n📊 Distribuição de Números Primos:")
print("-" * 70)

primos = {2, 3, 5, 7, 11, 13, 17, 19, 23}
primos_por_combinacao = []

for combo in combinacoes_validas:
    qtd_primos = sum(1 for n in combo if n in primos)
    primos_por_combinacao.append(qtd_primos)

distribuicao_primos = Counter(primos_por_combinacao)

for qtd_primos in sorted(distribuicao_primos.keys()):
    qtd = distribuicao_primos[qtd_primos]
    pct = (qtd / len(combinacoes_validas)) * 100
    bar = "█" * int(pct / 2)
    print(f"  {qtd_primos} primos: {qtd:6,} ({pct:5.2f}%) {bar}")

print(f"\n{'='*70}")


ESTATÍSTICAS DAS COMBINAÇÕES - SOMA 217

📊 Distribuição de Paridade:
----------------------------------------------------------------------
  4 pares / 11 ímpares:    767 ( 2.18%) █
  6 pares / 9 ímpares: 14,232 (40.46%) ████████████████████
  8 pares / 7 ímpares: 18,418 (52.36%) ██████████████████████████
  10 pares / 5 ímpares:  1,757 ( 4.99%) ██
  12 pares / 3 ímpares:      4 ( 0.01%) 

📊 Distribuição de Números Primos:
----------------------------------------------------------------------
  0 primos:      1 ( 0.00%) 
  1 primos:     24 ( 0.07%) 
  2 primos:    388 ( 1.10%) 
  3 primos:  2,462 ( 7.00%) ███
  4 primos:  7,502 (21.33%) ██████████
  5 primos: 11,686 (33.22%) ████████████████
  6 primos:  9,181 (26.10%) █████████████
  7 primos:  3,415 ( 9.71%) ████
  8 primos:    505 ( 1.44%) 
  9 primos:     14 ( 0.04%) 



In [158]:
if df_ultimos is not None:
    print(f"\n{'='*70}")
    print(f"ANÁLISE DOS ÚLTIMOS {ULTIMOS_N_JOGOS} JOGOS")
    print(f"{'='*70}\n")
    
    # Extrair números de cada jogo
    jogos_numeros = []
    somas_recentes = []
    
    for idx, row in df_ultimos.iterrows():
        # Tentar diferentes formatos de colunas
        numeros = []
        
        # Formato 1: colunas Bola1, Bola2, etc
        if 'Bola1' in df_ultimos.columns:
            numeros = [row[f'Bola{i}'] for i in range(1, 16)]
        # Formato 2: colunas numeradas 1, 2, 3, etc
        elif '1' in df_ultimos.columns:
            numeros = [row[str(i)] for i in range(1, 16)]
        # Formato 3: coluna 'Numeros' com string
        elif 'Numeros' in df_ultimos.columns:
            numeros_str = str(row['Numeros'])
            # Remover colchetes e split
            numeros = [int(n.strip()) for n in numeros_str.replace('[', '').replace(']', '').split(',')]
        
        if numeros:
            jogos_numeros.append(numeros)
            somas_recentes.append(sum(numeros))
    
    if somas_recentes:
        # Estatísticas de soma
        print("📊 ESTATÍSTICAS DE SOMA:")
        print("-" * 70)
        print(f"  Soma Mínima: {min(somas_recentes)}")
        print(f"  Soma Máxima: {max(somas_recentes)}")
        print(f"  Soma Média:  {sum(somas_recentes)/len(somas_recentes):.1f}")
        print(f"  Soma Alvo:   {SOMA_ALVO} {'✅ (dentro da faixa)' if min(somas_recentes) <= SOMA_ALVO <= max(somas_recentes) else '⚠️  (fora da faixa recente)'}")
        
        # Distribuição de somas
        print(f"\n📊 DISTRIBUIÇÃO DE SOMAS (últimos {ULTIMOS_N_JOGOS} jogos):")
        print("-" * 70)
        
        dist_somas = Counter(somas_recentes)
        for soma in sorted(dist_somas.keys()):
            count = dist_somas[soma]
            bar = "█" * count
            destaque = " ⬅️" if soma == SOMA_ALVO else ""
            print(f"  Soma {soma:3d}: {bar} ({count}x){destaque}")
        
        # Análise de paridade nos últimos jogos
        print(f"\n📊 PARIDADE NOS ÚLTIMOS {ULTIMOS_N_JOGOS} JOGOS:")
        print("-" * 70)
        
        paridades_recentes = []
        for numeros in jogos_numeros:
            pares = sum(1 for n in numeros if n % 2 == 0)
            paridades_recentes.append(pares)
        
        dist_paridade = Counter(paridades_recentes)
        for pares in sorted(dist_paridade.keys()):
            count = dist_paridade[pares]
            impares = 15 - pares
            pct = (count / len(paridades_recentes)) * 100
            bar = "█" * int(pct / 5)
            destaque = " ⬅️" if pares == 8 else ""
            print(f"  {pares} pares / {impares} ímpares: {count:2d}x ({pct:5.1f}%) {bar}{destaque}")
        
        # Análise de composição nos últimos jogos
        print(f"\n📊 COMPOSIÇÃO MÉDIA (últimos {ULTIMOS_N_JOGOS} jogos):")
        print("-" * 70)
        
        grupos = {
            'c1': list(range(1, 6)),
            'c2': list(range(6, 11)),
            'c3': list(range(11, 16)),
            'c4': list(range(16, 21)),
            'c5': list(range(21, 26))
        }
        
        composicoes = {g: [] for g in grupos.keys()}
        
        for numeros in jogos_numeros:
            for grupo, nums_grupo in grupos.items():
                count = sum(1 for n in numeros if n in nums_grupo)
                composicoes[grupo].append(count)
        
        print(f"  Grupo  Faixa   Média  Mín  Máx  Alvo")
        print("-" * 70)
        composicao_alvo = [4, 2, 3, 2, 4]
        for i, (grupo, valores) in enumerate(composicoes.items()):
            media = sum(valores) / len(valores)
            minimo = min(valores)
            maximo = max(valores)
            faixa = f"{grupos[grupo][0]:2d}-{grupos[grupo][-1]:2d}"
            alvo_str = f"{composicao_alvo[i]}"
            print(f"  {grupo}     {faixa}    {media:4.1f}   {minimo}    {maximo}    {alvo_str}")
        
        print(f"\n{'='*70}")
    else:
        print("⚠️  Não foi possível extrair números dos jogos recentes.")
        print("   Verifique o formato do arquivo lotofacil_sorteios.csv")
else:
    print("⚠️  Análise histórica não disponível (arquivo não encontrado)")


ANÁLISE DOS ÚLTIMOS 25 JOGOS

📊 ESTATÍSTICAS DE SOMA:
----------------------------------------------------------------------
  Soma Mínima: 166
  Soma Máxima: 244
  Soma Média:  193.3
  Soma Alvo:   217 ✅ (dentro da faixa)

📊 DISTRIBUIÇÃO DE SOMAS (últimos 25 jogos):
----------------------------------------------------------------------
  Soma 166: █ (1x)
  Soma 170: █ (1x)
  Soma 171: █ (1x)
  Soma 182: █ (1x)
  Soma 183: █ (1x)
  Soma 184: ██ (2x)
  Soma 185: █ (1x)
  Soma 186: ██ (2x)
  Soma 189: █ (1x)
  Soma 190: █ (1x)
  Soma 192: ██ (2x)
  Soma 193: █ (1x)
  Soma 195: █ (1x)
  Soma 197: █ (1x)
  Soma 199: █ (1x)
  Soma 203: ███ (3x)
  Soma 210: █ (1x)
  Soma 211: █ (1x)
  Soma 214: █ (1x)
  Soma 244: █ (1x)

📊 PARIDADE NOS ÚLTIMOS 25 JOGOS:
----------------------------------------------------------------------
  5 pares / 10 ímpares:  2x (  8.0%) █
  6 pares / 9 ímpares:  3x ( 12.0%) ██
  7 pares / 8 ímpares:  5x ( 20.0%) ████
  8 pares / 7 ímpares:  7x ( 28.0%) █████ ⬅️
  9 pa

## 6. Análise dos Últimos Jogos

Analisar padrões nos últimos jogos para validar a soma escolhida:

## 6. Visualizar Amostra Aleatória

Exibir 20 combinações aleatórias do arquivo gerado:

In [159]:
import random

if len(combinacoes_validas) > 20:
    amostra = random.sample(combinacoes_validas, 20)
else:
    amostra = combinacoes_validas

print(f"\n{'='*70}")
print(f"AMOSTRA DE 20 COMBINAÇÕES ALEATÓRIAS")
print(f"{'='*70}\n")

for i, combo in enumerate(amostra, 1):
    combo_str = '-'.join([f'{num:02d}' for num in combo])
    pares = sum(1 for n in combo if n % 2 == 0)
    impares = 15 - pares
    print(f"{i:2d}. {combo_str}  (P:{pares} I:{impares})")

print(f"\n{'='*70}")


AMOSTRA DE 20 COMBINAÇÕES ALEATÓRIAS

 1. 01-04-07-09-10-14-15-16-17-18-19-20-21-22-24  (P:8 I:7)
 2. 01-03-06-08-09-10-11-16-18-20-21-22-23-24-25  (P:8 I:7)
 3. 05-06-07-08-09-10-12-15-17-18-19-20-22-24-25  (P:8 I:7)
 4. 02-04-05-06-09-11-13-16-18-19-20-22-23-24-25  (P:8 I:7)
 5. 01-02-05-06-09-14-16-17-18-19-20-21-22-23-24  (P:8 I:7)
 6. 01-02-05-08-09-14-15-16-17-19-20-21-22-23-25  (P:6 I:9)
 7. 02-05-07-09-11-12-13-14-15-18-19-20-23-24-25  (P:6 I:9)
 8. 01-05-07-08-09-12-14-15-16-18-19-21-23-24-25  (P:6 I:9)
 9. 01-02-03-06-10-12-13-16-19-20-21-22-23-24-25  (P:8 I:7)
10. 01-02-04-05-09-11-16-17-18-19-21-22-23-24-25  (P:6 I:9)
11. 03-04-05-07-11-12-13-15-16-19-20-21-22-24-25  (P:6 I:9)
12. 02-06-07-08-10-11-13-16-17-18-19-20-22-23-25  (P:8 I:7)
13. 04-06-07-09-10-11-12-13-17-18-20-21-22-23-24  (P:8 I:7)
14. 01-02-05-06-07-11-14-17-19-20-21-22-23-24-25  (P:6 I:9)
15. 02-03-05-07-08-12-13-15-18-19-21-22-23-24-25  (P:6 I:9)
16. 01-02-03-07-13-14-15-16-17-19-20-21-22-23-24  (P:6 I:9)
1

---

## 💡 Instruções de Uso

1. **Alterar a soma**: Modifique a variável `SOMA_ALVO` na célula de configuração
2. **Executar células**: Execute as células em ordem (ou todas de uma vez)
3. **Arquivo gerado**: O CSV será salvo em `data/combinacoes_soma_{SOMA_ALVO}.csv`

### Exemplos de somas comuns na Lotofácil:
- **Média histórica**: ~195
- **Faixa ideal**: 177 a 213
- **Mais comum**: 190-200

## 7. Filtrar Combinações por Composição Específica (4-2-3-2-4)

Filtrar as combinações geradas para manter apenas aquelas com a composição exata 4-2-3-2-4:

In [160]:
# Definir grupos de números
grupos = {
    'c1': list(range(1, 6)),     # 1-5
    'c2': list(range(6, 11)),    # 6-10
    'c3': list(range(11, 16)),   # 11-15
    'c4': list(range(16, 21)),   # 16-20
    'c5': list(range(21, 26))    # 21-25
}

# Lista de composições para tentar (ordem de prioridade)
composicoes_para_tentar = [
    (4, 2, 3, 2, 4),  # Composição principal
    (4, 3, 3, 2, 3),  # Alternativa 1
    (3, 3, 3, 3, 3),  # Distribuição equilibrada
    (4, 2, 4, 2, 3),  # Alternativa 2
    (3, 2, 4, 3, 3),  # Alternativa 3
    (4, 3, 2, 3, 3),  # Alternativa 4
]

print("="*70)
print(f"FILTRANDO COMBINAÇÕES - SOMA {SOMA_ALVO}")
print("="*70)

# Tentar cada composição até encontrar resultados
composicao_encontrada = None
combinacoes_filtradas = []

for composicao_alvo in composicoes_para_tentar:
    print(f"\n🔍 Testando composição {composicao_alvo}...")
    
    temp_combinacoes = []
    
    for combo in combinacoes_validas:
        # Contar por grupo
        c1 = sum(1 for n in combo if n in grupos['c1'])
        c2 = sum(1 for n in combo if n in grupos['c2'])
        c3 = sum(1 for n in combo if n in grupos['c3'])
        c4 = sum(1 for n in combo if n in grupos['c4'])
        c5 = sum(1 for n in combo if n in grupos['c5'])
        
        composicao = (c1, c2, c3, c4, c5)
        
        # Verificar se corresponde à composição alvo
        if composicao == composicao_alvo:
            combo_str = '-'.join([f'{num:02d}' for num in combo])
            temp_combinacoes.append({
                'ID': len(temp_combinacoes) + 1,
                'Soma': SOMA_ALVO,
                'Combinacao': combo_str,
                'Numeros': str(list(combo)),
                'c1': c1,
                'c2': c2,
                'c3': c3,
                'c4': c4,
                'c5': c5
            })
    
    if len(temp_combinacoes) > 0:
        combinacoes_filtradas = temp_combinacoes
        composicao_encontrada = composicao_alvo
        print(f"   ✅ Encontradas {len(temp_combinacoes):,} combinações!")
        break
    else:
        print(f"   ⚠️  Nenhuma combinação encontrada.")

# Criar DataFrame
df_filtrado = pd.DataFrame(combinacoes_filtradas)

if composicao_encontrada:
    # Salvar arquivo
    comp_str = '_'.join(map(str, composicao_encontrada))
    arquivo_filtrado = data_dir / f'combinacoes_soma_{SOMA_ALVO}_composicao_{comp_str}.csv'
    df_filtrado.to_csv(arquivo_filtrado, index=False, encoding='utf-8-sig')
    
    print(f"\n{'='*70}")
    print(f"✅ COMPOSIÇÃO SELECIONADA: {composicao_encontrada}")
    print(f"{'='*70}")
    print(f"  c1 (1-5):   {composicao_encontrada[0]} números")
    print(f"  c2 (6-10):  {composicao_encontrada[1]} números")
    print(f"  c3 (11-15): {composicao_encontrada[2]} números")
    print(f"  c4 (16-20): {composicao_encontrada[3]} números")
    print(f"  c5 (21-25): {composicao_encontrada[4]} números")
    print(f"\n✅ Combinações filtradas: {len(combinacoes_filtradas):,}")
    print(f"   De: {len(combinacoes_validas):,} combinações originais")
    print(f"   Percentual: {(len(combinacoes_filtradas)/len(combinacoes_validas)*100):.2f}%")
    print(f"\n📁 Arquivo salvo: {arquivo_filtrado}")
else:
    composicao_alvo = composicoes_para_tentar[0]  # Manter a primeira para referência
    arquivo_filtrado = data_dir / f'combinacoes_soma_{SOMA_ALVO}_sem_composicao.csv'
    df_filtrado.to_csv(arquivo_filtrado, index=False, encoding='utf-8-sig')
    print(f"\n⚠️  Nenhuma composição encontrou resultados!")
    print(f"   Nenhuma das {len(composicoes_para_tentar)} composições testadas retornou combinações.")

print("="*70)

FILTRANDO COMBINAÇÕES - SOMA 217

🔍 Testando composição (4, 2, 3, 2, 4)...
   ⚠️  Nenhuma combinação encontrada.

🔍 Testando composição (4, 3, 3, 2, 3)...
   ⚠️  Nenhuma combinação encontrada.

🔍 Testando composição (3, 3, 3, 3, 3)...
   ⚠️  Nenhuma combinação encontrada.

🔍 Testando composição (4, 2, 4, 2, 3)...
   ⚠️  Nenhuma combinação encontrada.

🔍 Testando composição (3, 2, 4, 3, 3)...
   ⚠️  Nenhuma combinação encontrada.

🔍 Testando composição (4, 3, 2, 3, 3)...
   ⚠️  Nenhuma combinação encontrada.

⚠️  Nenhuma composição encontrou resultados!
   Nenhuma das 6 composições testadas retornou combinações.


## 8. Preview das Combinações Filtradas

In [161]:
# Visualizar as primeiras combinações filtradas
print(f"Total de combinações filtradas: {len(df_filtrado):,}")

if len(df_filtrado) > 0:
    # Usar a composição que foi realmente encontrada
    comp_usada = composicao_encontrada if composicao_encontrada else composicao_alvo
    print(f"\nPrimeiras 10 combinações com composição {comp_usada}:\n")
    print(df_filtrado.head(10).to_string(index=False))

    # Estatísticas das composições
    print(f"\n{'='*80}")
    print("Estatísticas das Composições:")
    print(f"{'='*80}")
    for col in ['c1', 'c2', 'c3', 'c4', 'c5']:
        print(f"{col}: média={df_filtrado[col].mean():.2f}, "
              f"min={df_filtrado[col].min()}, "
              f"max={df_filtrado[col].max()}")

    # Distribuição de paridade nas combinações filtradas
    pares_dist = df_filtrado['Combinacao'].apply(
        lambda x: sum(1 for n in x.split('-') if int(n) % 2 == 0)
    ).value_counts().sort_index()

    print(f"\n{'='*80}")
    print("Distribuição de Paridade (Números Pares):")
    print(f"{'='*80}")
    for pares, count in pares_dist.items():
        impares = 15 - pares
        pct = (count / len(df_filtrado)) * 100
        print(f"{pares} pares / {impares} ímpares: {count:,} combinações ({pct:.1f}%)")
else:
    print("\n⚠️  Nenhuma combinação encontrada com nenhuma das composições testadas!")
    print("   A soma escolhida não permite as distribuições testadas.")
    print(f"\n💡 Dica: Tente usar uma soma diferente.")

Total de combinações filtradas: 0

⚠️  Nenhuma combinação encontrada com nenhuma das composições testadas!
   A soma escolhida não permite as distribuições testadas.

💡 Dica: Tente usar uma soma diferente.


## 9. Filtrar por Paridade (8 Pares / 7 Ímpares)

In [162]:
# Filtrar combinações com 8 pares e 7 ímpares
print("="*70)
print(f"FILTRANDO PARIDADE - 8 PARES / 7 ÍMPARES")
print("="*70)

combinacoes_paridade = []

for index, row in df_filtrado.iterrows():
    # Extrair números da string de combinação
    numeros = [int(n) for n in row['Combinacao'].split('-')]
    
    # Contar pares e ímpares
    qtd_pares = sum(1 for n in numeros if n % 2 == 0)
    qtd_impares = 15 - qtd_pares
    
    # Verificar se é 8 pares / 7 ímpares
    if qtd_pares == 8 and qtd_impares == 7:
        combinacoes_paridade.append({
            'ID': len(combinacoes_paridade) + 1,
            'Soma': row['Soma'],
            'Combinacao': row['Combinacao'],
            'Numeros': row['Numeros'],
            'c1': row['c1'],
            'c2': row['c2'],
            'c3': row['c3'],
            'c4': row['c4'],
            'c5': row['c5'],
            'Pares': qtd_pares,
            'Impares': qtd_impares
        })

# Criar DataFrame
df_paridade = pd.DataFrame(combinacoes_paridade)

# Salvar arquivo
arquivo_paridade = data_dir / f'combinacoes_soma_{SOMA_ALVO}_comp_4_2_3_2_4_8p_7i.csv'
df_paridade.to_csv(arquivo_paridade, index=False, encoding='utf-8-sig')

print(f"\n✅ Combinações com 8 pares / 7 ímpares: {len(combinacoes_paridade):,}")
print(f"   De: {len(df_filtrado):,} combinações com composição 4-2-3-2-4")

if len(combinacoes_paridade) > 0:
    print(f"   Percentual: {(len(combinacoes_paridade)/len(df_filtrado)*100):.2f}%")
    print(f"\n📁 Arquivo salvo: {arquivo_paridade}")
    
    # Mostrar primeiras 10 combinações
    print(f"\n{'='*70}")
    print("PRIMEIRAS 10 COMBINAÇÕES:")
    print(f"{'='*70}")
    print(df_paridade.head(10)[['ID', 'Combinacao', 'c1', 'c2', 'c3', 'c4', 'c5']].to_string(index=False))
    
    # Análise de distribuição de pares por grupo
    print(f"\n{'='*70}")
    print("ANÁLISE DE PARES POR GRUPO:")
    print(f"{'='*70}")
    
    # Para cada grupo, contar quantos números pares existem
    grupos_info = {
        'c1': [2, 4],           # números pares em c1 (1-5)
        'c2': [6, 8, 10],       # números pares em c2 (6-10)
        'c3': [12, 14],         # números pares em c3 (11-15)
        'c4': [16, 18, 20],     # números pares em c4 (16-20)
        'c5': [22, 24]          # números pares em c5 (21-25)
    }
    
    for grupo, pares_grupo in grupos_info.items():
        total_numeros = df_paridade[grupo].iloc[0]  # quantidade do grupo na composição
        print(f"{grupo} (qtd: {total_numeros}): {len(pares_grupo)} números pares disponíveis -> {pares_grupo}")
    
    print("="*70)
else:
    print(f"\n⚠️  Nenhuma combinação encontrada com 8 pares e 7 ímpares!")
    
    if len(df_filtrado) > 0:
        print(f"\nDistribuição de paridade nas {len(df_filtrado):,} combinações:")
        pares_dist = df_filtrado['Combinacao'].apply(
            lambda x: sum(1 for n in x.split('-') if int(n) % 2 == 0)
        ).value_counts().sort_index()
        
        for pares, count in pares_dist.items():
            impares = 15 - pares
            pct = (count / len(df_filtrado)) * 100
            print(f"  {pares} pares / {impares} ímpares: {count:,} combinações ({pct:.1f}%)")
    else:
        print(f"\n💡 Não há combinações com a composição 4-2-3-2-4 para mostrar a distribuição.")
    
    print("="*70)

FILTRANDO PARIDADE - 8 PARES / 7 ÍMPARES

✅ Combinações com 8 pares / 7 ímpares: 0
   De: 0 combinações com composição 4-2-3-2-4

⚠️  Nenhuma combinação encontrada com 8 pares e 7 ímpares!

💡 Não há combinações com a composição 4-2-3-2-4 para mostrar a distribuição.


## 10. Filtrar por Similaridade (9 Números Iguais / 6 Diferentes)

In [163]:
# Buscar combinações com 9 números iguais e 6 diferentes
# Usar a primeira combinação encontrada para a soma como referência
if len(combinacoes_validas) > 0:
    referencia = list(combinacoes_validas[0])
else:
    referencia = [1, 2, 3, 4, 6, 7, 11, 14, 15, 18, 20, 22, 23, 24, 25]

print("="*70)
print("BUSCANDO COMBINAÇÕES SIMILARES - 9 IGUAIS / 6 DIFERENTES")
print("="*70)
print(f"\n🎯 Soma trabalhada: {SOMA_ALVO}")
print(f"\nCombinação de referência (1ª combinação da soma {SOMA_ALVO}):")
print(f"  {'-'.join([f'{n:02d}' for n in referencia])}")

# Verificar qual DataFrame usar como base
if len(combinacoes_paridade) > 0:
    df_base = df_paridade
    origem = "combinações com 8 pares / 7 ímpares"
else:
    df_base = df_filtrado
    origem = "combinações com composição 4-2-3-2-4"

print(f"\nBuscando em: {len(df_base):,} {origem}")

# Filtrar combinações com 9 números em comum
combinacoes_similares = []

for index, row in df_base.iterrows():
    # Extrair números da string de combinação
    numeros = [int(n) for n in row['Combinacao'].split('-')]
    
    # Calcular interseção
    set_referencia = set(referencia)
    set_numeros = set(numeros)
    
    iguais = set_referencia & set_numeros
    saem = set_referencia - set_numeros
    entram = set_numeros - set_referencia
    
    # Verificar se tem exatamente 9 números iguais
    if len(iguais) == 9:
        combinacoes_similares.append({
            'ID': len(combinacoes_similares) + 1,
            'Combinacao': row['Combinacao'],
            'Soma': row['Soma'],
            'c1': row['c1'],
            'c2': row['c2'],
            'c3': row['c3'],
            'c4': row['c4'],
            'c5': row['c5'],
            'Pares': row.get('Pares', sum(1 for n in numeros if n % 2 == 0)),
            'Impares': row.get('Impares', sum(1 for n in numeros if n % 2 == 1)),
            'Iguais': len(iguais),
            'Saem': sorted(saem),
            'Entram': sorted(entram)
        })

print(f"\n✅ Encontradas: {len(combinacoes_similares):,} combinações com 9 números iguais")

if len(combinacoes_similares) > 0:
    # Criar DataFrame
    df_similares = pd.DataFrame(combinacoes_similares)
    
    # Salvar arquivo
    arquivo_similares = data_dir / f'combinacoes_soma_{SOMA_ALVO}_9iguais_6diferentes.csv'
    df_similares.to_csv(arquivo_similares, index=False, encoding='utf-8-sig')
    print(f"📁 Arquivo salvo: {arquivo_similares}")
    
    # Mostrar primeiras 10 combinações
    print(f"\n{'='*70}")
    print("PRIMEIRAS 10 COMBINAÇÕES SIMILARES:")
    print(f"{'='*70}")
    
    for idx, row in df_similares.head(10).iterrows():
        print(f"\n{row['ID']:2d}. {row['Combinacao']}")
        print(f"    Composição: c1={row['c1']} c2={row['c2']} c3={row['c3']} c4={row['c4']} c5={row['c5']}")
        print(f"    Paridade: {row['Pares']}p / {row['Impares']}i")
        print(f"    Saem:   {row['Saem']}")
        print(f"    Entram: {row['Entram']}")
    
    print(f"\n{'='*70}")
    print(f"Total de combinações similares: {len(df_similares):,}")
    print("="*70)
else:
    print("\n⚠️  Nenhuma combinação encontrada com 9 números iguais à referência!")
    print("="*70)

BUSCANDO COMBINAÇÕES SIMILARES - 9 IGUAIS / 6 DIFERENTES

🎯 Soma trabalhada: 217

Combinação de referência (1ª combinação da soma 217):
  01-02-03-04-05-13-17-18-19-20-21-22-23-24-25

Buscando em: 0 combinações com composição 4-2-3-2-4

✅ Encontradas: 0 combinações com 9 números iguais

⚠️  Nenhuma combinação encontrada com 9 números iguais à referência!
